### PR TITLE
Implement Buffer Protocol

### PR DESCRIPTION
**Pull request by qingshi163**
I add a slot for implement the buffer protocol, more or less like what CPython doing. I am not sure is the right way for RustPython. Also I am confusing about the slots seem do not inherit from the base? So how we control the behavour of the subclass?

### CODE REVIEW

**Comment by youknowone** [comment]
Slots is a data field of type (PyClass). For now, we typically iterate mro to find proper field of slots.

File: vm/src/obj/objmemory.rs
Comment (coolreader18) [inline vm/src/obj/objmemory.rs:16]: Instead of `Send + Sync` here, you probably want to use `PyThreadingConstraint` from `pyobject`

**Comment by youknowone** [comment]
related to #2125 and #2195, right?

**Comment by qingshi163** [comment]
#2195 I had done it, I will push out when I passed the memoryview unittest.
#2125 I don't think it works with buffer, we have to do something inside objbytes.

**Comment by youknowone** [comment]
@qingshi163 I attached a commit to change Box to plain fn https://github.com/RustPython/RustPython/pull/2226/commits/c0ecc35ea5b76fe37edcb0e8d5d170ef1c759a17

File: vm/src/slots.rs
Comment (youknowone) [inline vm/src/slots.rs:386]: I think the original name `AsBuffer` you used was better. "hashable object" is a term in python(or cpython?) but bufferable is not.

Though "comparable object" is also nothing in python. @coolreader18 any idea about naming?

Comment (coolreader18) [reply vm/src/slots.rs:386]: I think AsBuffer is good, "buffer" isn't a verb so it doesn't really make sense as `Verbable` and there's precedence for `AsNoun` as a trait name in Rust, e.g. `AsRawFd`, `AsRef`

Comment (qingshi163) [reply vm/src/slots.rs:386]: AsBuffer is good, but the trait function name I changed it to get_buffer as same as CPython, also because it did return a new BufferProtocol. so we should go back to AsBuffer or something like GetBuffer? what we perfer?

File: vm/src/obj/objmemory.rs
Comment (qingshi163) [inline vm/src/obj/objmemory.rs:47]: the consumer should call as_contiguous if they want just a plain buffer. Should we also need to add a function here for consuming a non-contiguous buffer like a sliced memoryview? Like a iter() return the unpacked object?

File: vm/src/obj/objmemory.rs
Comment (qingshi163) [inline vm/src/obj/objmemory.rs:30]: I renamed as_bytes to obj_bytes because it should always return the full memory range for the original object. as_bytes may easier to be miss using.

File: vm/src/stdlib/pystruct.rs
Comment (qingshi163) [inline vm/src/stdlib/pystruct.rs:66]: I have to change the unit size for 'l' and 'L' to 8, because that is how array.array doing. I don't know is any compatible problem, but in my machine CPython 'l' and 'L' is 8 bytes.

Comment (youknowone) [reply vm/src/stdlib/pystruct.rs:66]: isn't it architecture dependent value?

Comment (qingshi163) [reply vm/src/stdlib/pystruct.rs:66]:
```
    /* Integers */
    case 'h':
        intsize = sizeof(short);
        is_signed = 1;
        break;
    case 'H':
        intsize = sizeof(short);
        is_signed = 0;
        break;
    case 'i':
        intsize = sizeof(int);
        is_signed = 1;
        break;
    case 'I':
        intsize = sizeof(int);
        is_signed = 0;
        break;
    case 'l':
        intsize = sizeof(long);
        is_signed = 1;
        break;
    case 'L':
        intsize = sizeof(long);
        is_signed = 0;
        break;
    case 'q':
        intsize = sizeof(long long);
        is_signed = 1;
        break;
    case 'Q':
        intsize = sizeof(long long);
        is_signed = 0;
        break;
```
This is from CPython, should we put it somewhere together? because we have to make sure array.array and struct have the same size for the type.

Comment (youknowone) [reply vm/src/stdlib/pystruct.rs:66]: For common platforms, using std::mem::size_of::<isize> or cfg target_pointer_width work. It fits for common platforms but not perfect. using libc always will fit as it defined in CPython. Try `std::mem::size_of::<libc::c_int>`

File: vm/src/slots.rs
Comment (coolreader18) [reply vm/src/slots.rs:386]: I personally think we can keep it as `AsBuffer`

File: vm/src/stdlib/pystruct.rs
Comment (qingshi163) [reply vm/src/stdlib/pystruct.rs:66]: I think we should do it in another PR and try unify the reflect from format code to size and the type.

File: common/src/borrow.rs
Comment (coolreader18) [inline common/src/borrow.rs:30]: This isn't sound -- somebody could match to get the mutex guard out of the enum, and then treat it as mutable. I ran into the same problem, and I'm not sure exactly what the solution is. Maybe try something in cell? ImmutableMappedMutexGuard or something?

Comment (coolreader18) [reply common/src/borrow.rs:30]: I could try making that later today if you don't feel confident about it; I think it would be like `{ data: *const T, raw: &'a parking_lot::RawMutex, _marker: PhantomData<(&'a T, RawMutex::Guard market)> }`

Comment (qingshi163) [reply common/src/borrow.rs:30]: If you can fix that will be perfect, I have no clue to get it soundless..

File: vm/src/slots.rs
Comment (youknowone) [reply vm/src/slots.rs]: That sounds like the implementation trait name can be `Buffer` and the desriptor trait name can be `BufferProtocol`. Like,
- AsBuffer(Bufferable) -> BufferProtocol
- BufferProtocol -> Buffer

Comment (coolreader18) [reply vm/src/slots.rs]: That makes sense, since a trait is inherently a "protocol", so there's no need to specify that stuff that a buffer can do is a protocol; it's redundant

File: common/src/borrow.rs
Comment (coolreader18) [reply common/src/borrow.rs:30]: #2239

File: vm/src/slots.rs
Comment (qingshi163) [reply vm/src/slots.rs]: That sounds good!

**Comment by qingshi163** [comment]
But the obj and the buffer in the memoryview could be different, the obj should always point to the original data source and the buffer is the controller shows how can we visit the data. That is how I implement it as the memoryview build from a memoryview the obj is cloned from original object.

**Comment by qingshi163** [comment]
I want to know if it is ok to merge, it is not fully completed but works.

**Comment by coolreader18** [comment]
Oh, I suppose that's true. Nevermind

File: Lib/test/test_memoryview.py
Comment (youknowone) [inline Lib/test/test_memoryview.py:115]: I know this PR enabled a lot of parts from this test, but unless it is fully resolved, I prefer to keep the whole test as expectedFailure before merge. this is only trackable by code, but expectedFailure is trackable by test result. Once the feature is fully done, it will be detected if it is expectedFailiure but not if it is commented out.

File: extra_tests/snippets/memoryview.py
Comment (youknowone) [inline extra_tests/snippets/memoryview.py:9]: is this intended?

File: vm/src/bytesinner.rs
Comment (qingshi163) [inline vm/src/bytesinner.rs:270]: Now we have different behavour between (memoryview op bytes) and (bytes op memoryview), Is any solution better than check the type here?

File: common/src/borrow.rs
Comment (qingshi163) [reply common/src/borrow.rs:30]: @coolreader18 will you merge #2239 before or after the pr?

File: vm/src/bytesinner.rs
Comment (qingshi163) [inline vm/src/bytesinner.rs:54]: I think we could have a trait for the objects that can be borrowed as &[PyObjectRef], so we can have a generic and efficiency way to iter.

File: vm/src/bytesinner.rs
Comment (coolreader18) [reply vm/src/bytesinner.rs:270]: Does memory view not implement the buffer protocol? I think that would maybe fix the issue(?)

Comment (qingshi163) [reply vm/src/bytesinner.rs:270]: But memoryview is implemented buffer protocol, we will have much more issue if not so.

Comment (coolreader18) [reply vm/src/bytesinner.rs:270]: Huh, that's strange, I think that's probably fine for now.

**Comment by qingshi163** [comment]
What that random ci fail comes from?

**Comment by coolreader18** [comment]
@qingshi163 I rebased this to master with the new mutex/`BorrowedValue::map` implementations; it put me as a committer for all the commits but hopefully you can just rebase it yourself and it would get rid of that.

**Comment by qingshi163** [comment]
@coolreader18 Thanks, I think I need review for merge now.

**Review by coolreader18 — APPROVED** [review]
I think this overall looks really good!! I'm fine with merging it and fixing leftover issues later, since this is so big already. @youknowone what do you think?

**Comment by coolreader18** [comment]
And don't worry about that CI failure; I can fix that when I merge.

**Comment by youknowone** [comment]
oops, after a few days of resting, I totally forgot to merge this first before other PRs. I agree to merge this and fix other stuffs later. Let me fix the conflicts.

**Comment by youknowone** [comment]
I fixed a few commit messages because fisrt two 'fix test' commits are not fixing test but disabling it.
I also reordered commits a bit to merge "update memoryview eq"s to be single commit and also "Implement buffer protocol" and "Rename to AsBuffer"

**Pull request merged**
